### PR TITLE
Refactor api_endpoint to be more explicit

### PR DESCRIPTION
**Pull request by claudijd**
# Description:

The input validation checks on SRV api endpoint discovery are sane, but it's usage afterwards could be better.  This PR incorporates feedback from david@bamsoftware.com, to make this even safer and more explicit how we're handling this potentially unsafe input that helps with API discovery.
______________

# Tasks:

- [X] Describe the problem / feature
- [X] Write tests
- [X] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems/remote_fetcher.rb [hunk 1]
```diff
@@ -109,9 +109,12 @@ def api_endpoint(uri)
       uri
     else
       target = res.target.to_s.strip
+      target_host = URI("https://" + target).host
 
-      if URI("http://" + target).host.end_with?(".#{host}")
-        return URI.parse "#{uri.scheme}://#{target}#{uri.path}"
```
Comment (claudijd) [inline lib/rubygems/remote_fetcher.rb]: Note that previously we incorporated the unparsed target content without it being host trimmed.  The this PR simply makes this much more explicit and, to me, it feels safer.

File: lib/rubygems/remote_fetcher.rb [hunk 2]
```diff
@@ -109,9 +109,12 @@ def api_endpoint(uri)
...
+      if target_host.end_with?(".#{host}")
+        uri = uri.dup
+        uri.host = target_host
+        return uri
```
Comment (segiddins) [inline lib/rubygems/remote_fetcher.rb]: this `return uri` is redundant it's already returned 3 lines below

Comment (claudijd) [reply lib/rubygems/remote_fetcher.rb]: Now that i think about it more, I'm thinking the assignment of the #dup call should be something other than uri, like modified_uri, to make it more clear that we're returning something other than the original uri.  That probably makes this more relevant and less likely to stop the uri variable name in this context.

**Comment by refashioned** [comment]
It occurs to me that a better/simpler way to handle this would be to call [URI::Generic.build](https://ruby-doc.org/stdlib-2.4.3/libdoc/uri/rdoc/URI/Generic.html#method-c-build) to build the URI, rather than string interpolation. Roughly, replace this:
```
      if URI("http://" + target).host.end_with?(".#{host}")
        return URI.parse "#{uri.scheme}://#{target}#{uri.path}"
      end
```
with this (plus error handling):
```
      if target.end_with?(".#{host}")
        return URI::Generic.build :scheme=>uri.scheme, :host=>target, :path=>uri.path
      end
```
Alternatively, call `URI::Generic.build` first, and then compare its `.host` against `".#{host}"`.

`URI::Generic.build` ensures that there's no interference between different components of the URI. `target` can only go into the host component, and if it has any bytes that would make that impossible, `URI::Generic.build` raises an error.

This example program shows `URI::Generic.build` rejecting all sorts of fishy targets which were the reason for opening this pull request. Note that the current code accepts all the dodgy targets.
```
require 'uri'
scheme = "https"
path = "/path"
[
  "example.com",
  "example.com:9999",
  "user:pass@example.com",
  "example.com/bad?q=",
  "example.com/bad#frag",
].each do |target|
  uri1 = URI.parse("#{scheme}://#{target}#{path}").to_s
  begin
    uri2 = URI::Generic.build(:scheme=>scheme, :host=>target, :path=>path).to_s
  rescue URI::InvalidComponentError => err
    uri2 = "##{err.class}"
  end
  puts "%-24s  %-34s  %s" % [target.inspect, uri1, uri2]
end
```
`target`                 | string interpolation               | `URI::Generic.build`
-------------------------|------------------------------------|----------------------------
"example.com"            | https://example.com/path           | https://example.com/path
"example.com:9999"       | https://example.com:9999/path      | #URI::InvalidComponentError
"user:pass@example.com"  | https://user:pass@example.com/path | #URI::InvalidComponentError
"example.com/bad?q="     | https://example.com/bad?q=/path    | #URI::InvalidComponentError
"example.com/bad#frag"   | https://example.com/bad#frag/path  | #URI::InvalidComponentError

P.S. if someone can make https://hackerone.com/reports/274267 public, it explains more about the rationale for this change.

**Comment by claudijd** [comment]
@refashioned I'll request the bug be publicly disclosed, these changes are non-exploitable to our knowledge, but more an improvement on the existing pattern to be more parnoid in how we handle these.  Please also note that we need to consider the contexts of CVE-2015-3900, CVE-2015-4020, CVE-2017-0902 to avoid a regression here, which is effectively an RCE outcome if we get it wrong.

**Comment by refashioned** [comment]
#2418 would remove `api_endpoint` completely and make this pull request moot. IMO #2418 is better than this pull request, for security anyway.

**Comment by claudijd** [comment]
I agree with @refashioned that removing the dynamic capability is most certainly safer/healthier IMO.  I'm going to close this PR in favor of #2418, this mimics the ask we made in 2015, when we found and disclosed CVE-2015-3900 that started this saga.

**Pull request closed without merging**